transport_plugins/jlink/iotile_transport_jlink/jlink.py

```diff
@@ -163,166 +160,82 @@ def _try_connect(self, connection_string):
+            'write_memory': self._jlink_async.debug_write_memory,
+            'program_flash': self._jlink_async.program_flash
         }
 
-        cmd_code = known_commands.get(cmd_name)
-        if cmd_code is None:
-            callback(conn_id, self.id, False, None, "Unsupported command: %s" % cmd_name)
+        self._ensure_connection(conn_id, True)
 
-        def _on_finished(_name, retval, exception):
-            if exception is not None:
-                callback(conn_id, self.id, False, None, str(exception))
-                return
+        func = known_commands.get(name)
+        if name is None:
+            raise ArgumentError("Unsupported command: %s" % name)
 
-            callback(conn_id, self.id, True, retval, None)
+        return await func(self._device_info, self._control_info, cmd_args)
 
-        self._control_thread.command(cmd_code, _on_finished, self._device_info, self._control_info, cmd_args, progress_callback)
 
-    def connect_async(self, connection_id, connection_string, callback):
+    async def connect(self, conn_id, connection_string):
         """Connect to a device by its connection_string
 
-        This function asynchronously connects to a device by its BLE address
-        passed in the connection_string parameter and calls callback when
-        finished.  Callback is called on either success or failure with the
-        signature:
-
-        callback(conection_id, adapter_id, success: bool, failure_reason: string or None)
-
-        Args:
-            connection_string (string): A unique connection string that identifies
-                which device to connect to, if many are possible.
-            connection_id (int): A unique integer set by the caller for
-                referring to this connection once created
-            callback (callable): A callback function called when the
-                connection has succeeded or failed
+        This function asynchronously connects to a device by device type.
+        See iotile_transport_jlink_devices.py
         """
-        self._try_connect(connection_string)
+        await self._try_connect(connection_string)
+        self._setup_connection(conn_id, connection_string)
 
-        def _on_finished(_name, control_info, exception):
-            if exception is not None:
-                callback(connection_id, self.id, False, str(exception))
-                return
+        try:
+            self._control_info = await self._jlink_async.verify_control_structure(self._device_info, self._control_info)
+        except HardwareError as e:
+            logger.exception("RPC, streaming and tracing won't work, but device still can be flashed or debugged")
 
-            if control_info is not None:
-                self._control_info = control_info
+        self._connection_id = conn_id
 
-            callback(connection_id, self.id, True, None)
-            self._connection_id = connection_id
+    async def open_interface(self, conn_id, interface):
```

Comment:
I don't believe this will work as implemented.  There are multiple interfaces that can be opened and closed independently:

- `rpc`, `tracing`, `streaming`, `debug`, `script`

`debug` and `scrip` open/close may be no-ops for jlink and `rpc`, `tracing`, `streaming` may have the same underlying `poll` method feeding data to each, however:
- `trace` events must only be generated when `tracing` interface is open
- `report` events must only be generated when `report` interface is open

Also, the status of the `trace` and `streaming`  interfaces needs to be communicated to the device so that it can implement flow control.  For example, inside the controller the `stream_manager` subsystem needs to know whether the `streaming` interface is open/closed because it calls the `open_stream` rpc and that must fail if no jlink adapter is connected and no one is listening for tracing/report data.

So, it seems like you need two different things here:

1. If any of `rpc`, `tracing`, `streaming`  are open, then `polling` is on
2. A separate flag for whether each individual interface is opened.  The `tracing` and `streaming` interface flags need to be communicated to the firmware so that it knows if it can send tracing/streaming data.  

Additionally, stale data needs to be cleared (if present) on opening `tracing`/`streaming` interfaces if it's possible for stale data to be stuck in the fifo.

Reply:
I will add flags as you suggest